File: backend/app/schemas/redeem.py
from datetime import datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class RedeemEligibilityResponse(BaseModel):
    market_id: int
    condition_id: str
    winning_outcome: str | None
    eligible: bool
    status: str
    reasons: list[str] = Field(default_factory=list)
    real_winning_order_exists: bool = False
    matched_winning_size: Decimal | None = None


class RedeemRecordResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    market_id: int
    settlement_id: int | None
    condition_id: str
    winning_outcome: str
    status: str
    mode: str
    tx_hash: str | None
    wallet_address: str | None
    amount_redeemed: Decimal | None
    balance_before: Decimal | None
    balance_after: Decimal | None
    error_message: str | None
    raw_response: dict[str, Any]
    created_at: datetime
    updated_at: datetime


class RedeemAttemptResult(BaseModel):
    market_id: int
    condition_id: str
    winning_outcome: str
    status: str
    record: RedeemRecordResponse | None = None
    tx_hash: str | None = None
    amount_redeemed: Decimal | None = None
    balance_before: Decimal | None = None
    balance_after: Decimal | None = None
    error_message: str | None = None
    reasons: list[str] = Field(default_factory=list)


class RedeemStatusResponse(BaseModel):
    market_id: int
    condition_id: str
    winning_outcome: str | None
    status: str
    tx_hash: str | None = None
    amount_redeemed: Decimal | None = None
    balance_before: Decimal | None = None
    balance_after: Decimal | None = None
    error_message: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
    real_winning_order_exists: bool = False
    reasons: list[str] = Field(default_factory=list)

File: frontend/types/redeem.ts
export type RedeemStatus =
  | "NOT_ELIGIBLE"
  | "READY_TO_REDEEM"
  | "REDEEM_SUBMITTED"
  | "REDEEM_CONFIRMED"
  | "REDEEM_FAILED"
  | "SKIPPED_DRY_RUN"
  | "SKIPPED_PAPER_ONLY";

export interface RedeemRecord {
  id: number;
  market_id: number;
  settlement_id: number | null;
  condition_id: string;
  winning_outcome: string;
  status: RedeemStatus;
  mode: string;
  tx_hash: string | null;
  wallet_address: string | null;
  amount_redeemed: string | null;
  balance_before: string | null;
  balance_after: string | null;
  error_message: string | null;
  raw_response: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface RedeemStatusResponse {
  market_id: number;
  condition_id: string;
  winning_outcome: string | null;
  status: RedeemStatus;
  tx_hash: string | null;
  amount_redeemed: string | null;
  balance_before: string | null;
  balance_after: string | null;
  error_message: string | null;
  created_at: string | null;
  updated_at: string | null;
  real_winning_order_exists: boolean;
  reasons: string[];
}

export interface RedeemAttemptResult {
  market_id: number;
  condition_id: string;
  winning_outcome: string;
  status: RedeemStatus;
  record: RedeemRecord | null;
  tx_hash: string | null;
  amount_redeemed: string | null;
  balance_before: string | null;
  balance_after: string | null;
  error_message: string | null;
  reasons: string[];
}
